File: UserDrivers/TCP/tcp_server.c
/**
 * MIT License
 *
 * Copyright (c) 2025 ElectronicsBuilder
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 * @file    tcp_server.c
 * @brief   TCP Server Task Implementation
 * @details FreeRTOS task implementation for TCP server operations.
 *          Manages WiFi TCP server lifecycle, client connections, and status monitoring.
 */

#include "tcp_server.h"
#include "main.h"
#include "log.hpp"

#include "wifi_tcpClient.h"
#include "wifi_tcpServer.h"
#include "main_app.h"

// ============================================================================
// Private Function Declarations
// ============================================================================

static void waitWifiInit(void);

// ============================================================================
// External Variables
// ============================================================================

extern bool TouchGFX_init;
extern bool wifiDriverInit;
extern bool PeripheralsTestComplete;
// ============================================================================
// TCP Server Task Implementation
// ============================================================================

/**
 * @brief TCP Server FreeRTOS task
 * @details Main task function that manages TCP server operations:
 *          1. Waits for TouchGFX initialization
 *          2. Waits for WiFi driver initialization
 *          3. Initializes TCP server on default port (8080)
 *          4. Starts TCP server
 *          5. Processes client connections in main loop
 *          6. Logs server status every 30 seconds
 * @param argument Task argument (unused)
 */
void tcpServerTask(void *argument)
{
    osDelay(6000);
    // Wait for TouchGFX to be initialized
    while (!TouchGFX_init && !PeripheralsTestComplete) {
        osDelay(100);
    }
    main_app();
    // Wait for Wi-Fi to be initialized
    waitWifiInit();

    LOG_INFO("[TCP Server] Task started, initializing server...");

    // Initialize TCP server on default port (8080)
    if (wifi_tcp_server_init(0) != 0) {
        LOG_ERROR("[TCP Server] Failed to initialize server");
        goto error_exit;
    }

    // Start TCP server
    if (wifi_tcp_server_start() != 0) {
        LOG_ERROR("[TCP Server] Failed to start server");
        goto error_exit;
    }

    LOG_INFO("[TCP Server] Server started successfully");

    // Main server loop
     for (;;) {
        // Process server operations
        int32_t clients_served = wifi_tcp_server_process();

        if (clients_served < 0) {
            LOG_ERROR("[TCP Server] Error processing server operations");
            break;
        }

        // Give other tasks a chance to run
        osDelay(1);  // Reduced from 100ms to 10ms for better responsiveness

        // Periodically log server status
        static uint32_t last_status_log = 0;
        uint32_t now = HAL_GetTick();
        if (now - last_status_log > 30000) { // Every 30 seconds
            tcp_server_state_t state;
            uint32_t active_clients, total_connections;
            wifi_tcp_server_get_status(&state, &active_clients, &total_connections);

            LOG_INFO("[TCP Server] Status: State=%d, Active=%lu, Total=%lu",
                     state, active_clients, total_connections);
            last_status_log = now;
        }

     }

error_exit:
    LOG_ERROR("[TCP Server] Task exiting due to error");
    wifi_tcp_server_stop();

    // Task should not exit, but if it does, suspend itself
    for (;;) {
        osDelay(10000);
     }
}

// ============================================================================
// Private Helper Functions
// ============================================================================

/**
 * @brief Wait for WiFi driver initialization
 * @details Blocks until wifiDriverInit flag is set to true.
 *          Polls every 1 second to check initialization status.
 */
static void waitWifiInit(void)
{
    do {
        osDelay(1000);
    } while (wifiDriverInit != true);
}

File: tools/tcp_server/tcp_server.py
#!/usr/bin/env python3
"""
/**
 * MIT License
 * 
 * Copyright (c) 2025 ElectronicsBuilder
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 * 
 * @file    tcp_server.py
 * @brief   TCP Server for STM32 FalkraDrones Wi-Fi Testing
 */
"""

import socket
import threading
import time
import sys

class TCPServer:
    def __init__(self, host='0.0.0.0', port=8080):
        self.host = host
        self.port = port
        self.server_socket = None
        self.running = False

    def handle_client(self, client_socket, client_address):
        """Handle individual client connections"""
        print(f"[{time.strftime('%H:%M:%S')}] New connection from {client_address}")

        try:
            while True:
                # Receive data from client
                data = client_socket.recv(1024)
                if not data:
                    break

                message = data.decode('utf-8', errors='ignore')
                print(f"[{time.strftime('%H:%M:%S')}] Received: {repr(message)}")

                # Send response back to client
                response = f"Server received: {message}"
                client_socket.send(response.encode('utf-8'))
                print(f"[{time.strftime('%H:%M:%S')}] Sent response: {repr(response)}")

        except Exception as e:
            print(f"[{time.strftime('%H:%M:%S')}] Error handling client {client_address}: {e}")
        finally:
            client_socket.close()
            print(f"[{time.strftime('%H:%M:%S')}] Connection closed: {client_address}")

    def start(self):
        """Start the TCP server"""
        try:
            self.server_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            self.server_socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
            self.server_socket.settimeout(1.0)  # Add timeout to make accept() non-blocking
            self.server_socket.bind((self.host, self.port))
            self.server_socket.listen(5)
            self.running = True

            print(f"[{time.strftime('%H:%M:%S')}] TCP Server started on {self.host}:{self.port}")
            print(f"[{time.strftime('%H:%M:%S')}] Waiting for STM32 connections...")
            print("Press Ctrl+C to stop the server")

            while self.running:
                try:
                    client_socket, client_address = self.server_socket.accept()

                    # Handle each client in a separate thread
                    client_thread = threading.Thread(
                        target=self.handle_client,
                        args=(client_socket, client_address)
                    )
                    client_thread.daemon = True
                    client_thread.start()

                except socket.timeout:
                    # Timeout allows checking self.running periodically
                    continue
                except socket.error as e:
                    if self.running:
                        print(f"[{time.strftime('%H:%M:%S')}] Socket error: {e}")
                        break

        except KeyboardInterrupt:
            print(f"\n[{time.strftime('%H:%M:%S')}] Received Ctrl+C, shutting down...")
        except Exception as e:
            print(f"[{time.strftime('%H:%M:%S')}] Server error: {e}")
        finally:
            self.stop()

    def stop(self):
        """Stop the TCP server"""
        self.running = False
        if self.server_socket:
            self.server_socket.close()
        print(f"\n[{time.strftime('%H:%M:%S')}] Server stopped")

def main():
    """Main function"""
    # Default settings
    host = '0.0.0.0'  # Listen on all interfaces
    port = 8080

    # Parse command line arguments
    if len(sys.argv) > 1:
        port = int(sys.argv[1])
    if len(sys.argv) > 2:
        host = sys.argv[2]

    # Create and start server
    server = TCPServer(host, port)

    try:
        server.start()
    except KeyboardInterrupt:
        print(f"\n[{time.strftime('%H:%M:%S')}] Received Ctrl+C, shutting down...")
        server.stop()

if __name__ == '__main__':
    main()